pkg/cli/email/validate.go

```diff
@@ -0,0 +1,154 @@
+// It runs checks in the order of cheapest to most expensive.
+func Validate(emailAddress string, options ...Option) error {
+	if _, ok := disposableDomains[emailAddress]; ok {
+		return fmt.Errorf(`disposable address "%s" not allowed`, emailAddress)
+	}
+
+	cfg := &optConfig{}
+	for _, applyOption := range options {
+		applyOption(cfg)
+	}
+
+	if cfg.withRFC522 {
+		found := validRfc5322Regexp.Find([]byte(emailAddress))
+
+		if string(found) != emailAddress {
+			return errors.New("not RFC522 compliant")
+		}
+	}
+
+	// Split the email into two parts: local and domain, separated by '@'
+	parts := strings.Split(emailAddress, "@")
+	if len(parts) != 2 {
+		return errors.New("missing @")
+	}
+
+	localPart, domainPart := parts[0], parts[1]
+
+	// Check if both the local part and domain part are non-empty
+	if len(localPart) == 0 || len(domainPart) == 0 {
+		return errors.New("missing parts")
+	}
+
+	// Validate the domain part contains a dot and has text on both sides
+	if strings.Count(domainPart, ".") == 0 || strings.HasPrefix(domainPart, ".") || strings.HasSuffix(domainPart, ".") {
+		return errors.New(`missing "."`)
+	}
+
+	// Ensure the domain has at least one character between dots
+	domainLabels := strings.Split(domainPart, ".")
+	for _, label := range domainLabels {
+		if len(label) == 0 {
+			return fmt.Errorf(`double dots in "%s"`, emailAddress)
+		}
+	}
+
+	// Check if the local part doesn't start or end with a dot
+	if strings.HasPrefix(localPart, ".") || strings.HasSuffix(localPart, ".") {
+		return fmt.Errorf(`cannot have "." at the beginning or end of "%s"`, emailAddress)
+	}
+
+	// Check that the local part doesn't contain consecutive dots
+	if strings.Contains(localPart, "..") {
+		return fmt.Errorf(`double dots in "%s"`, localPart)
+	}
+
+	if cfg.checkMX {
+		return CheckMXRecords(domainPart, cfg.checkMXTimeout)
+	}
+
+	return nil
+}
+
+func WithRFC522() Option {
+	return func(c *optConfig) {
+		c.withRFC522 = true
+	}
+}
+
+func WithCheckMX() Option {
+	return func(c *optConfig) {
+		c.checkMX = true
+		c.checkMXTimeout = time.Second
+	}
+}
+
+func WithCheckMXTimeout(duration time.Duration) Option {
+	return func(c *optConfig) {
+		c.checkMX = true
+		c.checkMXTimeout = duration
+	}
+}
+
+// CheckMXRecords tries to make a connection with one of the MX records for the given domain
+// times out after the given duration.
+func CheckMXRecords(domain string, timeout time.Duration) error {
+	records, err := net.LookupMX(domain)
+	if err != nil {
+		return err
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+	errc := make(chan error)
+
+	// Fire off a go routine for each record and SMTP port
+	// The first to return a nil error will return success
+	for _, record := range records {
+		for _, port := range smtpPorts {
+			go func() {
+				_, connectError := smtp.Dial(fmt.Sprintf("%s:%d", record.Host, port))
+				errc <- connectError
```

Comment:
Couldn't this lead to the other go routines being blocked by trying to send to errc, which isn't read anymore in case of a canceled context or the first nil err received?

Reply:
Good catch, updated.